src/Interpreters/ExpressionActions.cpp

```diff
@@ -1702,70 +1024,250 @@ std::string ActionsDAG::dumpNames() const
     return out.str();
 }
 
-ExpressionActionsPtr ActionsDAG::buildExpressions(const Context & context)
+void ActionsDAG::removeUnusedActions(const Names & required_names)
 {
-    struct Data
+    std::unordered_set<Node *> nodes_set;
+    std::vector<Node *> required_nodes;
+    required_nodes.reserve(required_names.size());
+
+    for (const auto & name : required_names)
     {
-        Node * node = nullptr;
-        size_t num_created_children = 0;
-        size_t num_expected_children = 0;
-        std::vector<Node *> parents;
-        Node * renamed_child = nullptr;
-    };
+        auto it = index.find(name);
+        if (it == index.end())
+            throw Exception(ErrorCodes::UNKNOWN_IDENTIFIER,
+                            "Unknown column: {}, there are only columns {}", name, dumpNames());
 
-    std::vector<Data> data(nodes.size());
-    std::unordered_map<Node *, size_t> reverse_index;
+        if (nodes_set.insert(*it).second)
+            required_nodes.push_back(*it);
+    }
 
-    for (auto & node : nodes)
+    removeUnusedActions(required_nodes);
+}
+
+void ActionsDAG::removeUnusedActions(const std::vector<Node *> & required_nodes)
+{
     {
-        size_t id = reverse_index.size();
-        data[id].node = &node;
-        reverse_index[&node] = id;
+        Index new_index;
+
+        for (auto * node : required_nodes)
+            new_index.insert(node);
+
+        index.swap(new_index);
     }
 
-    std::queue<Node *> ready_nodes;
-    std::queue<Node *> ready_array_joins;
+    removeUnusedActions();
+}
 
-    for (auto & node : nodes)
+void ActionsDAG::removeUnusedActions()
+{
+    std::unordered_set<const Node *> visited_nodes;
+    std::stack<Node *> stack;
+
+    for (auto * node : index)
     {
-        data[reverse_index[&node]].num_expected_children += node.children.size();
+        visited_nodes.insert(node);
+        stack.push(node);
+    }
 
-        for (const auto & child : node.children)
-            data[reverse_index[child]].parents.emplace_back(&node);
+    while (!stack.empty())
+    {
+        auto * node = stack.top();
+        stack.pop();
 
-        if (node.renaming_parent)
+        if (!node->children.empty() && node->column && isColumnConst(*node->column) && node->allow_constant_folding)
         {
+            /// Constant folding.
+            node->type = ActionsDAG::Type::COLUMN;
+            node->children.clear();
+        }
 
-            auto & cur = data[reverse_index[node.renaming_parent]];
-            cur.renamed_child = &node;
-            cur.num_expected_children += 1;
+        for (auto * child : node->children)
+        {
+            if (visited_nodes.count(child) == 0)
+            {
+                stack.push(child);
+                visited_nodes.insert(child);
+            }
         }
     }
 
-    for (auto & node : nodes)
+    nodes.remove_if([&](const Node & node) { return visited_nodes.count(&node) == 0; });
```

Comment:
Seems like we need to erase removed nodes:
```
 A call to remove is typically followed by a call to a container's erase method, which erases the unspecified values and reduces the physical size of the container to match its new logical size. 
```

Reply:
It is remove for `std::list`
```
The similarly-named container member functions list::remove, list::remove_if, forward_list::remove, and forward_list::remove_if erase the removed elements.
```